[DO NOT MERGE] Raven-DB-11605 - For review only

grisha-kotler opened the pull request.

grisha-kotler committed:
RavenDB-11605 - created the MongoDB migration tool

grisha-kotler committed:
RavenDB-11605 - added CosmosDB migrator

grisha-kotler committed:
RavenDB-11605 - add a controller that will be used by the studio,
add DatabaseMigration operation

grisha-kotler committed:
RavenDB-11605 - revert change to DatabaseDestination

grisha-kotler committed:
RavenDB-11605 - studio changes: added basic migration options

talweiss1982 commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,131 @@ public Task GetMigratedServerUrls()
+                    {
+                        var line = process.StandardOutput.ReadLine();
+                        // return type is json, will throw if we fail to deserialize
+                        JsonConvert.DeserializeObject(line);
```
This seems a bit harsh to deserialize the JSON just for validation, I'm sure there are cheaper ways to achieve this.

ayende replied:
I don't like it either, this is just throwing a lot of perf out the window

ayende replied:
Better to pass it to the client and have the error there

talweiss1982 commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,131 @@ public Task GetMigratedServerUrls()
+                                using (ContextPool.AllocateOperationContext(out DocumentsOperationContext migrateContext))
+                                {
+                                    var options = new DatabaseSmugglerOptionsServerSide();
+                                    var stream = new GZipStream(process.StandardOutput.BaseStream, CompressionMode.Decompress);
```
This needs to be disposed of.

grisha-kotler replied:
done

talweiss1982 commented on tools/Raven.Migrator/CollectionsInfo.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Raven.Migrator
+{
+    public class CollectionsInfo
```
Is this class expected to have more members in the future? (seems too much to have a class for holding a single primitive member)

grisha-kotler replied:
It is rarely used and is easier to convert to JSON this way

talweiss1982 commented on tools/Raven.Migrator/CosmosDB/CosmosDBMigrator.cs:
```diff
@@ -0,0 +1,169 @@
+                        document,
+                        documentId,
+                        ravenCollectionName,
+                        new List<string>
```
You're generating this const list per document, it would be better to define this as read-only static list

grisha-kotler replied:
done

talweiss1982 commented on tools/Raven.Migrator/MigrationHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+        {
+            if (configuration.ConsoleExport)
+            {
+                var jsonString = JsonConvert.SerializeObject(obj);
```
Maybe we can use EntityToBlittable here and use Sparrow to save memory (just performance enhancements)

grisha-kotler replied:
can't use Sparrow in the Migrator

ppekrol replied:
disable indentation?

ppekrol replied:
and I think that you can add Raven.Client and Sparrow.dll if needed.

talweiss1982 commented on tools/Raven.Migrator/MongoDB/MongoDBMigrator.cs:
```diff
@@ -0,0 +1,273 @@
+                            document,
+                            documentId,
+                            ravenCollectionName,
+                            new List<string>
```
Better use a static read-only list here

grisha-kotler replied:
done

grisha-kotler committed:
RavenDB-11605 - PR comments

grisha-kotler committed:
RavenDB-11605 - allow to add a transform script

grisha-kotler committed:
RavenDB-11605 - fix studio validation

ml054 reviewed (commented):
I assume the progress of operation is same as in smuggler since we read output on migrator and using smuggler from import right?

ml054 reviewed (approved):
Studio changes looks good.

arekpalinski commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,135 @@ public Task GetMigratedServerUrls()
+                    }
+                    catch (Exception e)
+                    {
+                        var error = process.StandardError.ReadLine();
```
we can throw here as well I think

ayende replied:
You are only reading a single line, what if there are several?

arekpalinski commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,135 @@ public Task GetMigratedServerUrls()
+                                                        + processStartInfo.FileName + " " + processStartInfo.Arguments, e);
+                }
+
+                await process.StandardInput.WriteLineAsync(migrationConfiguration.Input.ToString());
```
what do we pass here? is it  "Command - available commands: databases, collections, export"?

grisha-kotler replied:
https://github.com/grisha-kotler/ravendb/blob/fd50c270b7f01149d8cd7d13af5e738c6be100b9/tools/Raven.Migrator/MongoDB/MongoDBConfiguration.cs

or

https://github.com/grisha-kotler/ravendb/blob/fd50c270b7f01149d8cd7d13af5e738c6be100b9/tools/Raven.Migrator/CosmosDB/CosmosDBConfiguration.cs#L3

grisha-kotler replied:
It can be any migration configuration

arekpalinski replied:
Thanks for clarification. `InputConfiguration` might be better property name then (minor)

grisha-kotler replied:
done

arekpalinski commented on tools/Raven.Migrator/MigrationHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+                await streamWriter.WriteAsync(",");
+            isFirstDocument.Value = false;
+
+            // we are going to recalculate the hash when importing
```
Sure? 
https://github.com/ravendb/ravendb/blob/v4.1/src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs#L571

grisha-kotler replied:
correct. need to calculate it here.

grisha-kotler replied:
done

grisha-kotler committed:
RavenDB-11605 - PR comments

ppekrol commented on src/Raven.Server/Documents/Operations/Operations.cs:
```diff
@@ -377,8 +377,8 @@ public enum OperationType
-            [Description("Database migration")]
-            DatabaseMigration,
+            [Description("RavenDB Database migration")]
+            DatabaseMigrationRavenDb,
```
`DatabaseMigrationFromRavenDB`

Not sure what implications does this rename have? Backward compatibility needed?

grisha-kotler replied:
Old migrations from RavenDB will display only the documents section (studio only)

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -299,7 +301,7 @@ private async Task BulkImport(BlockingCollection<Func<Task<Stream>>> files, stri
 
         [RavenAction("/databases/*/admin/smuggler/migrate", "POST", AuthorizationStatus.Operator)]
-        public async Task Migrate()
+        public async Task RavenDbMigration()
```
MigrateFromRavenDB

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,135 @@ public Task GetMigratedServerUrls()
+                        JsonConvert.DeserializeObject(line);
+                        using (var sw = new StreamWriter(ResponseBodyStream()))
+                        {
+                            await sw.WriteAsync(line);
```
What this `IsExportCommand == false` does? Where are we writing this single line and for what reason?

grisha-kotler replied:
The tool can return a JSON with a databases or collections list
or an export of the documents. I need a way to distinguish those two actions.

ppekrol replied:
IsExportCommand => ActionType (enum -> GetMetadata, Export) ?

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,135 @@ public Task GetMigratedServerUrls()
+                            }
+                            catch (Exception e)
+                            {
+                                var error = process.StandardError.ReadLine();
```
You need to read whole error here. Check how this is done in Embedded for example.

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Smuggler/Migration/MigrationConfiguration.cs:
```diff
@@ -1,8 +1,22 @@
 {
+    public class MigrationConfiguration
+    {
+        public string DatabaseTypeName { get; set; }
```
enum?

ayende replied:
Still waiting

grisha-kotler replied:
this is "mongodb" or "cosmosdb"

ppekrol replied:
But can be more in the future. Enum would be better?

ppekrol commented on src/Raven.Server/Smuggler/Migration/MigrationConfiguration.cs:
```diff
@@ -1,8 +1,22 @@
+    {
+        public string DatabaseTypeName { get; set; }
+
+        public string FullPathToMigrator { get; set; }
```
MigratorFullPath

ayende replied:
Still waiting

ppekrol commented on src/Raven.Server/Smuggler/Migration/MigrationConfiguration.cs:
```diff
@@ -1,8 +1,22 @@
+
+        public string FullPathToMigrator { get; set; }
+
+        public BlittableJsonReaderObject Input { get; set; }
```
What is hidden behind this?

ppekrol commented on src/Raven.Server/Smuggler/Migration/MigrationConfiguration.cs:
```diff
@@ -1,8 +1,22 @@
+
+        public BlittableJsonReaderObject Input { get; set; }
+
+        public bool IsExportCommand { get; set; }
```
What is the purpose of this?

ppekrol commented on tools/Raven.Migrator/CosmosDB/CosmosDBMigrator.cs:
```diff
@@ -0,0 +1,171 @@
+            var azureCollections = client.CreateDocumentCollectionQuery(database.SelfLink).ToList();
+            foreach (var azureCollection in azureCollections)
+            {
+                collectionsToMigrate.Add(azureCollection.Id, azureCollection.Id);
```
you can simply this using `client.CreateDocumentCollectionQuery(database.SelfLink).ToDictionary()`

ppekrol commented on tools/Raven.Migrator/CosmosDB/CosmosDBMigrator.cs:
```diff
@@ -0,0 +1,171 @@
+        private void AssertDatabaseName()
+        {
+            if (string.IsNullOrWhiteSpace(_configuration.DatabaseName))
+                throw new ArgumentException("DatabaseName cannot be null or empty");
```
ArgumentNullException?

ppekrol commented on tools/Raven.Migrator/MigrationHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+                return;
+            }
+
+            foreach (var prop in obj.GetType().GetProperties())
```
Do we need to handle the Fields?

ppekrol commented on tools/Raven.Migrator/MigrationHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+            }
+            else
+            {
+                dictionary["@metadata"] = new Dictionary<string, object>
```
You can simply this and previous dict and remove the redundancy,

ppekrol commented on tools/Raven.Migrator/Raven.Migrator.csproj:
```diff
@@ -0,0 +1,21 @@
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
we need to add runtime version here also

ppekrol commented on tools/Raven.Migrator/Raven.Migrator.csproj:
```diff
@@ -0,0 +1,21 @@
+
+  <ItemGroup>
+    <Reference Include="Microsoft.Extensions.CommandLineUtils">
+      <HintPath>..\..\..\..\Users\grisha.HRHINOS\.nuget\packages\microsoft.extensions.commandlineutils\1.1.1\lib\netstandard1.3\Microsoft.Extensions.CommandLineUtils.dll</HintPath>
```
This does not look valid.

grisha-kotler replied:
fixed

ppekrol commented on tools/Raven.Migrator/MigrationHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+
+            using (stream)
+            using (var gzipStream = new GZipStream(stream, CompressionMode.Compress, leaveOpen: true))
+            using (var streamWriter = new StreamWriter(gzipStream))
```
could be easier with JsonWriter?

grisha-kotler replied:
done

ppekrol commented on tools/Raven.Migrator/CommandLineApp.cs:
```diff
@@ -0,0 +1,118 @@
+            "   MigrateGridFS - (Optional) migrate GridFS that is associated with the provided database" + Environment.NewLine +
+            "   CollectionsToMigrate - (Optional) a dictionary of collections to rename during the export";
+
+        private static readonly string CosmosDbDescription =
```
I think that this MS command line supports multi-level commands and multi-level help? Talk with @gregolsky about that.

gregolsky replied:
Have a look in https://github.com/ravendb/ravendb/blob/v4.1/tools/rvn/CommandLineApp.cs

grisha-kotler replied:
I have a JSON over there. Not sure how this can work.

ppekrol replied:
Then I suggest using 'nameof'

ppekrol commented on tools/Raven.Migrator/AbstractMigrationConfiguration.cs:
```diff
@@ -0,0 +1,17 @@
+
+        public bool ConsoleExport { get; set; }
+
+        public string ExportFilePath { get; set; }
```
`ConsoleExport = ExportFilePath == null` ? Can remove that one property?

grisha-kotler replied:
if ExportFilePath == null and ConsoleExport == true a file will be exported in the application folder

ppekrol replied:
Looks like a unnecessary side-effect?

grisha-kotler replied:
The tool can work as standalone and export a valid .ravendbdump file.
Raven.Migrator.exe export
{"Command": "export", "DatabaseName": "test" }

If I remove the ConsoleExport property this command will export to the console as ExportFilePath == null

ppekrol replied:
Can we export to two of those sources at once?

grisha-kotler replied:
no

ppekrol replied:
I think it would be better to throw when both of them are set or none of them is set?

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,135 @@ public Task GetMigratedServerUrls()
             return Task.CompletedTask;
         }
 
+        [RavenAction("/databases/*/smuggler/migrate", "POST", AuthorizationStatus.ValidUser)]
```
Admin operation?

grisha-kotler replied:
isn't it the same as an import?

ppekrol replied:
I think that it is the same as the other migration? `/databases/*/admin/smuggler/migrate`

BTW, since this is not exposed to the Client API, it would be better to change the
`/databases/*/admin/smuggler/migrate` -> `/databases/*/admin/smuggler/migrate/ravendb`

and this endpoint from `/databases/*/smuggler/migrate` -> `/databases/*/admin/smuggler/migrate`

ayende replied:
This *MUST* be an admin operation. 
We are allowing the user to specify an exec that we'll run. That means that we want to be pretty sure that only an authorized admin can do that, and not just any user.

ayende replied:
Also need to check that database restore from 3.x is admin only (which I think it is because this is creating the db)

grisha-kotler replied:
done

grisha-kotler replied:
DatabaseAdmin or Operator?

grisha-kotler replied:
changed to Operator as the smuggler/migrate/ravendb endpoint

grisha-kotler committed:
RavenDB-11605 - added collections to migrate to the studio

grisha-kotler committed:
RavenDB-11605 - add Microsoft.Extensions.CommandLineUtils nuget package

grisha-kotler committed:
RavenDB-11605 - Mןigrator: calculate the attachment hash in

grisha-kotler committed:
RavenDB-11605 - remove the IsExportCommand and check the InputConfiguration for the command type

grisha-kotler committed:
RavenDB-11605 - use JsonTextWriter when generating the export file

grisha-kotler committed:
RavenDB-11605 - read the error correctly

ayende commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,135 @@ public Task GetMigratedServerUrls()
+                            try
+                            {
+                                using (ContextPool.AllocateOperationContext(out DocumentsOperationContext migrateContext))
+                                using (var stream = new GZipStream(process.StandardOutput.BaseStream, CompressionMode.Decompress))
```
You are sending GZIP over the pipe? Why?
It is in process communication and the cost of compression is only wasting our time.

grisha-kotler replied:
done

ayende commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +383,135 @@ public Task GetMigratedServerUrls()
+                                var error = process.StandardError.ReadLine();
+                                result.AddError($"Error occurred during migration. Process error: {error}, exception: {e}");
+                                onProgress.Invoke(result.Progress);
+                                process.Kill();
```
Where are you closing the process when the operation completes normally?

grisha-kotler committed:
RavenDB-11605 - change migrate endpoints
/databases/*/admin/smuggler/migrate -> /databases/*/admin/smuggler/migrate/ravendb
/databases/*/smuggler/migrate -> /databases/*/admin/smuggler/migrate

grisha-kotler committed:
RavenDB-11605 - fix json string as return value

grisha-kotler committed:
RavenDB-11605 - fix checkbox toggle for collections

grisha-kotler committed:
RavenDB-11605 - fix endpoint that is used in the studio

grisha-kotler committed:
RavenDB-11605 - don't Gzip the smuggler output when it is in process communication

grisha-kotler committed:
RavenDB-11605 - migration is an Operator action

ayende commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +385,164 @@ public Task GetMigratedServerUrls()
+                {
+                    try
+                    {
+                        var line = process.StandardOutput.ReadLine();
```
Here you are reading a _single_ line, is that intentional, what about other output?

grisha-kotler replied:
Intentional. The ouput from tool is here: https://github.com/ravendb/ravendb/blob/1932a9d2799b886aad3040c4192624975fcb3148/tools/Raven.Migrator/MigrationHelpers.cs#L48

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +385,164 @@ public Task GetMigratedServerUrls()
+                }
+                catch (Exception e)
+                {
+                    process?.Kill();
```
I think we should wrap this in try/catch

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +385,164 @@ public Task GetMigratedServerUrls()
+                    catch (Exception e)
+                    {
+                        var errorString = await ReadOutput(process.StandardError).ConfigureAwait(false);
+                        process.Kill();
```
Wrap this in try/catch?

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +385,164 @@ public Task GetMigratedServerUrls()
+                            }
+                            catch (OperationCanceledException)
+                            {
+                                process.Kill();
```
wrap it in try/catch?

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +385,164 @@ public Task GetMigratedServerUrls()
+                            }
+                            catch (ObjectDisposedException)
+                            {
+                                process.Kill();
```
wrap it in try/catch?

ppekrol commented on src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:
```diff
@@ -381,6 +385,164 @@ public Task GetMigratedServerUrls()
+                                var errorString = await ReadOutput(process.StandardError).ConfigureAwait(false);
+                                result.AddError($"Error occurred during migration. Process error: {errorString}, exception: {e}");
+                                onProgress.Invoke(result.Progress);
+                                process.Kill();
```
...probably would be better to create a method for safely killing that?

ppekrol commented on tools/Raven.Migrator/CommandLineApp.cs:
```diff
@@ -0,0 +1,124 @@
+            "Connect to MongoDB, pass configuration as JSON" + Environment.NewLine +
+            "   Command - available commands: databases, collections, export" + Environment.NewLine +
+            "   ConnectionString - MongoDB connection string" + Environment.NewLine +
+            "   DatbaseName - the database name, applicaple only for the collections and export commands" + Environment.NewLine +
```
using 'nameof' here could be helpful, e.g. you could avoid this typo

grisha-kotler replied:
done

ppekrol commented on tools/Raven.Migrator/MigrationHelpers.cs:
```diff
@@ -0,0 +1,209 @@
+
+            if (attachments == null || attachments.Count == 0)
+            {
+                dictionary[MetadataProperty] = new Dictionary<string, object>
```
Wouldn't it be better to use this as a base and after that add 'if' condition to push the attachment entries?

grisha-kotler replied:
done

ppekrol commented on tools/Raven.Migrator/MongoDB/MongoDBMigrator.cs:
```diff
@@ -0,0 +1,264 @@
+                    foreach (var collectionDocument in cursor.Current)
+                    {
+                        var collectionName = collectionDocument["name"].ToString();
+                        if (collectionName.EndsWith(".chunks") || collectionName.EndsWith(".files"))
```
Maybe it would be better to ignore casing here?

grisha-kotler replied:
done

ppekrol commented on tools/Raven.Migrator/MongoDB/MongoDBMigrator.cs:
```diff
@@ -0,0 +1,264 @@
+            StreamWriter streamWriter)
+        {
+            var bucket = new GridFSBucket(database);
+            var collectionName = bucket.Options.BucketName == "fs" ? "Files" : bucket.Options.BucketName;
```
Can we have a casing issue here?

grisha-kotler replied:
The default name for GridFS is "fs".
If it was changed we'll take it as the collection name.

grisha-kotler committed:
RavenDB-11605 - PR comments

grisha-kotler committed:
RavenDB-11605 - add tooltips
